### PR TITLE
Add a Dockerfile that can build the collector along with an example.

### LINKED ISSUE #236: Add a Dockerfile that can build the binary too
Currently, it seems that there is a Dockerfile that can create an image from a binary built on the host machine. It would be nice to have a Dockerfile that can actually build the binary too, for example this one - https://github.com/anuraaga/opentelemetry-collector-contrib/blob/custom/Dockerfile

This way, it's easy to locally build a docker image for master without worrying about tooling using something like `docker build -t opentelemetry-collector:test https://github.com/opentelemetry/opentelemetry-collector-contrib.git` or similarly for local changes. This can work especially well with `docker-compose` which will always start up containers after building source without any other steps.

If this seems like a reasonable idea, I can send a PR with the Dockerfile I linked.

### PR DESCRIPTION
**Pull request by anuraaga**
**Description:** <Describe what has changed. 

Currently there is a Dockerfile which packages a prebuilt binary. It would be convenient to have a Dockerfile that actually builds the binary as well. This makes sure that go tooling isn't required to play with a binary built from master. It works particularly well with docker-compose, where you can set the build context to https://github.com/open-telemetry/opentelemetry-collector-contrib.git and always be testing integrations against master with a single command.

**Link to tracking Issue:** fixes #236 

**Testing:** 
docker build / docker run

**Documentation:** Added the docker build command to the README

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+RUN make otelcontribcol
+
+FROM alpine:latest as certs
+RUN apk --update add ca-certificates
+
+FROM scratch
+COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
+COPY --from=build /src/bin/otelcontribcol_linux_amd64 /otelcontribcol
+ENTRYPOINT ["/otelcontribcol"]
+EXPOSE 55678 55679
```
Comment (flands) [inline Dockerfile:16]: replace 55678 (opencensus) with 55680 (otlp)

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+RUN make otelcontribcol
+
+FROM alpine:latest as certs
+RUN apk --update add ca-certificates
+
+FROM scratch
+COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
+COPY --from=build /src/bin/otelcontribcol_linux_amd64 /otelcontribcol
+ENTRYPOINT ["/otelcontribcol"]
```
Comment (flands) [inline Dockerfile:15]: if a config is not specified then otel will not start

File: Dockerfile (hunk 1 above)
Comment (anuraaga) [reply Dockerfile:16]: The runtime package is copied from the other Dockerfile here

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/cmd/otelcontribcol/Dockerfile

Should I update both?

File: Dockerfile (hunk 2 above)
Comment (anuraaga) [reply Dockerfile:15]: Yeah - this is a package of the binary, are you thinking we should package a default config with it too? Seems like a nice idea if we know what a good default is.

**Comment by owais** [comment]
This can get a bit confusing. We now have two ways to build the docker image with the "non-standard" one being exposed on project root as Dockerfile making it very easy for people mistake it for the one that is used to publish real images. Another minor issue is that this project was laid out to host multiple binaries so we "bundled" docker file for each binary with it's own directory (cmd/otelcol for example). 

If the goal is to allow building using docker, can we just add a build image that just builds the binary and exits? The two docker files can be orthogonal that way. One used to build binaries and the other used to build docker images. Another thing I'd look into would be caching Go modules and build cache with the builder docker image.

**Comment by anuraaga** [comment]
Open to any suggestions to clean this up, indeed at the last moving to some subdirectory might make sense.

The goal of this was to make both building and running as seamless as possible. Someone can just do this and the binary will be automatically built and run.

`docker run -it --rm https://github.com/open-telemetry/opentelemetry-collector-contrib.git`

This works well with docker-compose. Unfortunately, having a Dockerfile that only builds a binary doesn't really help with this flow. The collector is usually used with many services to play with it, so docker-compose has been very helpful for me. If this doesn't seem so helpful for others, I'm fine with closing and maintaining my own forked repo with the Dockerfile, but thought this could come in handy.

File: Dockerfile (hunk 1 above)
Comment (flands) [reply Dockerfile:16]: Ah, yes good idea -- plan is to move from oc to ot

File: Dockerfile (hunk 2 above)
Comment (flands) [reply Dockerfile:15]: if the goal is to make it functional then a good idea -- examples directory has a config that can be used

**Comment by owais** [comment]
I suppose we could add a docker-compose directory that this build docker file and docker compose configuration as well. It could build and bring up the contrib collector will a standard/default configuration. Would be helpful for people trying out otel for the first time as well.

**Comment by anuraaga** [comment]
Thanks for the idea! I have moved into an examples directory and created a more complete example with a frontend, backend zipkin, and jaeger.

File: examples/tracing/otel-collector-config.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+receivers:
+  zipkin:
+    endpoint: 0.0.0.0:9411
```
Comment (owais) [inline examples/tracing/otel-collector-config.yml:3]: Suggest adding OTLP receiver as well even if example app doesn't yet export in OTLP

**Comment by anuraaga** [comment]
Sure added it

**Review by owais — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@owais is this Dockerfile verified/used during CircleCI build? I would not want it to just sit there unused and go stale.

**Comment by owais** [comment]
Makes sense @tigrannajaryan. We can add a CI job to run the example build. I guess we can just verify that the image builds as verifying the example works end to end might take a lot of time.

Alternatively, we can add a recurring job that runs every day and verifies the examples end to end.

**Comment by tigrannajaryan** [comment]
Just verofying that the build passes (exits with zero status code) is good enough for now.

**Comment by tigrannajaryan** [comment]
> Just verofying that the build passes (exits with zero status code) is good enough for now.

@anuraaga can you make the corresponding change to include the Docker build step in the CircleCI build?

**Comment by anuraaga** [comment]
Sorry for the noise, had to test by running CircleCI - set it up!

File: examples/tracing/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      - "16686:16686"
+      - "14268"
+      - "14250"
+
+  #Zipkin
+  zipkin:
+    image: openzipkin/zipkin
+    container_name: zipkin
+    ports:
+      - 9411:9411
```
Comment (owais) [inline examples/tracing/docker-compose.yml:16]: Do we need zipkin here considering jaeger all in one can accept zipkin and show the traces as well?

Comment (anuraaga) [reply examples/tracing/docker-compose.yml:16]: It's not required but it's neat to see OTel collector serving its vision to drive multiple backends, and it's helpful to be able to compare the views to make sure the exported data is sane. The non-contrib demo similarly starts them both https://github.com/open-telemetry/opentelemetry-collector/blob/master/examples/demo/docker-compose.yaml#L12

**Review by owais — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@anuraaga thank you!

**Pull request merged**